### PR TITLE
Support dynamic masks in splash attention

### PR DESCRIPTION
**Pull request by Rifur13**
Adding support for dynamic masks in the splash attention kernel.

Currently, splash attention expects a static mask. It's preprocessed, and only the *interesting* (not fully masked) parts of the mask are passed to the kernel. This change allows users to pass in a jax.Array instead. Since we can’t know the number of partial mask blocks at trace time, the entire mask is materialized in `partial_mask_blocks`.

### CODE REVIEW

File: tests/pallas/tpu_splash_attention_mask_test.py [hunk 1]
```diff
@@ -2091,6 +2092,89 @@ def test_process_invalid_mask(self):
...
+
+    def make_causal_mask(q_seq_len: int, kv_seq_len: int):
+      idx_q = jnp.arange(q_seq_len, dtype=np.int32)
+      idx_kv = jnp.arange(kv_seq_len, dtype=np.int32)
+
+      return idx_q[:, None] >= idx_kv[None, :]
+
+    mask = jnp.stack([make_causal_mask(q_seq_len, kv_seq_len)] * head_count)
+
+    mask_info, _ = mask_info_lib.process_dynamic_mask(
```
Comment (apghml) [inline tests/pallas/tpu_splash_attention_mask_test.py:2108]: Would it make sense to test this while running inside a jit context? IIUC, the current implementation of `downcast_to_small_dtype()` will not work under jit, causing the call to fail. A test that uses a jit context would catch this kind of issue.

File: jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py [hunk 2]
```diff
@@ -304,6 +315,134 @@ def _get_mask_info_for_shard(
...
+    data_next = _downcast_to_small_type(data_next)
+    block_mask = _downcast_to_small_type(block_mask)
+    mask_next = _downcast_to_small_type(mask_next)
+
+  return (
+      MaskInfo(
+          data_next=data_next,
+          mask_next=mask_next,
+          block_mask=block_mask,
+          partial_mask_blocks=partial_mask_blocks,
```
Comment (apghml) [inline jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py:439]: Does `partial_mask_blocks` get stored in scalar memory? What happens if we exceed the size of scalar memory? More generally, how would this handle e.g., a 100k+ length context where partial_mask_blocks is quite large. (Since we can't do deduplication for dynamic arrays.)

Some possible things that might help:
* Support int8 for partial_mask_blocks since its entries are all 0/1, or maybe even packed bool values.
* Have some sort of sharding support for partial_mask_blocks.
* Probably the most "comprehensive" solution would be to support `ComputableMask` having `jax.tree_util.Partial()` as the callable with the ability to specify sharding information for the arrays stored in the partial object. This would even allow users to implement support for the first bullet point themselves.

**Comment by apghml** [comment]
Thanks a lot!

File: jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py (hunk 2 above)
Comment (Rifur13) [reply jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py:439]: `partial_mask_blocks ` is in HBM, scalar memory is tiny on TPUs. Using `mask_next`, the right mask block is prefetched into VMEM for each kernel invocation. You still need to fit `partial_mask_blocks `  in HBM, so sharding and using smaller data types help here.

We're running into a known edge case of mosaic here by using int8/bools, but people are working on it. One workaround we can do for now is to use smaller data types and upcast to int32 later.

Thanks for the comments, let me think some more about sharding and get back to you.

File: tests/pallas/tpu_splash_attention_mask_test.py (hunk 1 above)
Comment (Rifur13) [reply tests/pallas/tpu_splash_attention_mask_test.py:2108]: Good catch, everything should work under jit here. I'll fix it.

File: jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py [hunk 3]
```diff
@@ -304,6 +308,136 @@ def _get_mask_info_for_shard(
...
+  is_full_mask = jnp.all(partial_mask_blocks, axis=(-1, -2))
+  is_empty_mask = jnp.logical_not(jnp.any(partial_mask_blocks, axis=(-1, -2)))
+
+  block_mask = jnp.ones(block_mask_shape, dtype=np.int32)
+  block_mask = jnp.where(is_full_mask, 2, block_mask)
+  block_mask = jnp.where(is_empty_mask, 0, block_mask)
+
+  mask_next = jnp.where(
+      jnp.logical_or(is_empty_mask, is_full_mask),
+      0,
```
Comment (sharadmv) [inline jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py:390]: Leave TODO/comment explaining choice of 0

**Review by sharadmv — APPROVED** [review]
Very nice!

File: jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py [hunk 4]
```diff
@@ -304,6 +308,136 @@ def _get_mask_info_for_shard(
...
+    if max_value <= np.iinfo(np.int8).max:
+      return array.astype(np.int8)
+    elif max_value <= np.iinfo(np.int16).max:
+      return array.astype(np.int16)
+    else:
+      return array.astype(np.int32)
+
+  if downcast_smem_data:
+    block_mask = block_mask.astype(np.int8)  # values are in the range [0, 1, 2]
+    data_next = _downcast(data_next, kv_seq_len if is_dkv else q_seq_len)
```
Comment (apghml) [inline jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py:426]: Should this be `kv_blocks_count` and `q_blocks_count` respectively?

File: jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py (hunk 2 above)
Comment (apghml) [reply jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py:434]: I just wanted to give a gentle ping about this.

File: jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py (hunk 4 above)
Comment (Rifur13) [reply jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py]: Good catch, done.

File: jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py [hunk 5]
```diff
@@ -304,6 +308,136 @@ def _get_mask_info_for_shard(
...
 
+def _process_dynamic_mask(
+    mask: jax.Array,
+    block_shape: tuple[int, int],
+    is_dkv: bool,
+    *,
+    downcast_smem_data: bool = True,
+    head_shards: int = 1,
+    q_seq_shards: int = 1,
+    shrink_grid: bool = True,
```
[the remark this answers is not in the document]
Comment (Rifur13) [reply jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py:319]: Done.

File: jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py [hunk 6]
```diff
@@ -304,6 +308,136 @@ def _get_mask_info_for_shard(
...
+      kv_blocks_count,
+  )
+
+  # Tile the last 2 dimensions of the mask into 2D tiles of size `block_shape`.
+  partial_mask_blocks = (
+      mask.reshape(head_count, q_blocks_count, q_block_size, kv_seq_len)
+      .swapaxes(-1, -2)
+      .reshape(*block_mask_shape, kv_block_size, q_block_size)
+      .swapaxes(-1, -2)
+      .astype(np.int32)
```
[the remark this answers is not in the document]
Comment (Rifur13) [reply jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py]: Rebased. The tests did indeed catch this.

File: jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py (hunk 3 above)
Comment (Rifur13) [reply jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_mask_info.py:404]: Done.
